tests/canary_upgrade_test.go

```diff
@@ -0,0 +1,189 @@
+		kv.Spec.CustomizeComponents = v1.CustomizeComponents{
+			Patches: []v1.CustomizeComponentsPatch{
+				{
+					ResourceName: "virt-handler",
+					ResourceType: "DaemonSet",
+					Type:         v1.StrategicMergePatchType,
+					Patch:        patch,
+				},
+			},
+		}
+		return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			_, err := virtCli.KubeVirt(flags.KubeVirtInstallNamespace).Update(kv)
+			return err
+		})
+	}
+
+	isVirtHandlerUpdated := func(ds *appsv1.DaemonSet) bool {
+		_, exists := ds.Spec.Template.Annotations[e2eCanaryTestAnnotation]
+		return exists
+	}
+
+	removeExpectedAtHead := func(eventsQueue []string, expectedEvent string) []string {
+		defer GinkgoRecover()
+
+		if expectedEvent == lastObservedEvent {
+			return eventsQueue
+		}
+		if len(eventsQueue) == 0 {
+			return eventsQueue
+		}
+
+		head := eventsQueue[0]
+		ExpectWithOffset(1, head).To(Equal(expectedEvent), fmt.Sprintf("was expecting event %s but got %s instead", expectedEvent, head))
+		lastObservedEvent = expectedEvent
+		return eventsQueue[1:]
+	}
+
+	processDsEvent := func(ds *appsv1.DaemonSet, eventsQueue []string) []string {
+		update := ds.Spec.UpdateStrategy.RollingUpdate
+		if update == nil {
+			return eventsQueue
+		}
+		maxUnavailable := update.MaxUnavailable
+		if maxUnavailable == nil {
+			return eventsQueue
+		}
+
+		if maxUnavailable.String() == "10%" {
+			eventsQueue = removeExpectedAtHead(eventsQueue, "maxUnavailable=10%")
+		}
+		if maxUnavailable.IntValue() == 1 {
+			eventsQueue = removeExpectedAtHead(eventsQueue, "maxUnavailable=1")
+		}
+		if ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+			pods, err := virtCli.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(context.Background(), metav1.ListOptions{LabelSelector: "kubevirt.io=virt-handler"})
+			Expect(err).ToNot(HaveOccurred())
+
+			var updatedPods int32
+			for i := range pods.Items {
+				if _, exists := pods.Items[i].Annotations[e2eCanaryTestAnnotation]; exists {
+					updatedPods++
+				}
+			}
+
+			if updatedPods > 0 && updatedPods == ds.Status.DesiredNumberScheduled {
+				eventsQueue = removeExpectedAtHead(eventsQueue, "virt-handler=ready")
+			}
+		}
+		return eventsQueue
+	}
+
+	It("should successfully upgrade virt-handler", func() {
+		expectedEvents := []string{
+			"maxUnavailable=1",
+			"maxUnavailable=10%",
+			"virt-handler=ready",
+			"maxUnavailable=1",
+		}
+
+		go dsInformer.Run(stopCh)
+		cache.WaitForCacheSync(stopCh, dsInformer.HasSynced)
+
+		dsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, curr interface{}) {
+				ds := curr.(*appsv1.DaemonSet)
+				expectedEvents = processDsEvent(ds, expectedEvents)
+			},
+		})
+
+		err := updateVirtHandler()
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() bool {
+			return len(expectedEvents) == 0
+		}, 120*time.Second, 1*time.Second).Should(BeTrue(), fmt.Sprintf("events %v were expected but did not occur", expectedEvents))
```

Comment:
the expectedEvents slice is being accessed by two goroutines at the same time without being protected by a mutex. This could cause issues since it's not technically thread safe.

Can you add a mutex around the `expectedEvents = processDsEvent(ds, expectedEvents)` and `len(expectedEvents) == 0`

Reply:
Thanks for spotting this, it is indeed something to protect with a mutex, already pushed the change.